### ISSUE TITLE
Likelihood fit returns no covariance matrix

### ISSUE THREAD

**Issue by egede** [tBuLi/symfit#138]
When I perform a  maximum likelihood fit, there is no covariance matrix returned. See example below

```from symfit import Parameter, Variable, Likelihood, exp
import numpy as np
import numpy as np
import matplotlib.pyplot as plt

# Define the model for a Gaussian
mu = Parameter(value=1.0)
sigma = Parameter(value=1.0, min=0.5)
x = Variable()
model = (1.0 / sigma) * exp(- (x-mu)*(x-mu) / (2.*sigma*sigma))

# Draw 1000 samples from a Gaussian
nsample=1000
data = np.random.normal(loc=5.0, scale=2.0, size=nsample)

# Do the fitting
fit = Likelihood(model, data)
fit_result = fit.execute(method='L-BFGS-B')
print(fit_result)
print(type(fit_result.params.covariance_matrix))``
```

results in 

```
Parameter Value        Standard Deviation
mu        4.936787e+00 None
sigma     2.000927e+00 None
Fitting status message: CONVERGENCE: NORM_OF_PROJECTED_GRADIENT_<=_PGTOL
Number of iterations:   18
Regression Coefficient: nan

<type 'NoneType'>
```
so minimum is found correctly but there is no covariance matrix. Is this a limitation of `symfit` or a problem with the code above?

**Comment by tBuLi** [comment]
Thanks for your question. Yes, this is known behavior, not a bug.

This is because calculating the parameter uncertainty from a log-likelihood fit is not (at first glance) the same as for a standard least-square type fit, so I simply havent found time yet to think about how to implement it correctly in this case. 

If you are willing to help out in this case it would be much appreciated, either with the mathematics or the coding. A source on how to do it is section 5 of http://www.sherrytowers.com/mle_introduction.pdf. I do think that if we use a minimizer which returns the Hessian of the optimization it might be the same procedure up to a minus sign but like I said, I should think about that more carefully ;).

Currently me and @pckroon are working on a mayor overhaul of most of the internals of `symfit` though, so if you do decide to contribute code it's best to do it against PR #133 .

**Comment by egede** [comment]
@tBuLi Thanks for getting back to me. Looking a bit more at the code, I see that if I use the method `L-BFGS-B` for the optimisation, then `ans.hess_inv.todense()` can be assigned as the covariance matrix. If I combine this with fitting for `-2 log-likelihood` (note the factor two), then this gives the correct result for a fit with a single parameter. For more than one parameter, I have not looked at if coverage is correct.

Below is an output of a test with the same code as above.

```
Parameter Value        Standard Deviation
mu        4.981034e+00 1.912356e-01
sigma     1.866041e+00 1.329029e-01
Fitting status message: CONVERGENCE: NORM_OF_PROJECTED_GRADIENT_<=_PGTOL
Number of iterations:   17
Regression Coefficient: nan

2*Log-likelihood at minimum: 2.247638e+02
2*Log-likelihood one stderr away: 2.258141e+02
Difference in 2*LL : 1.050256e+00
Covariance matrix
[[ 0.03657107  0.00031345]
 [ 0.00031345  0.01766319]]
```

I might indeed be able to contribute something to this.

**Comment by tBuLi** [comment]
@egede Thanks for that, that's very good to know. If you could find out with for a known toy model with two parameters you get the expected result with this method, that would be great.

If it holds for multiple parameters as well this will be very easy to implement.

**Comment by tBuLi** [comment]
I have looked into this further and have made some important findings.

As a test scenario, I extended the Tests.test_likelihood_fitting_gaussian in test_general.py. Here we sample N points from a normal distribution with known mean mu and standard deviation sigma. We then do a likelihood fit with a gaussian distribution to this sample and find back the known parameters mu and sigma:

```python
mu, sig = parameters('mu, sig')
sig.value = 3.0
mu.value = 50.0
N = 10000
x = Variable()
pdf = Gaussian(x, mu, sig)

np.random.seed(10)
xdata = np.random.normal(51., 3.5, N)

# Expected values
mean = np.mean(xdata)
stdev = np.std(xdata)
mean_stdev = stdev / np.sqrt(N)

print(mean, mean_stdev)
fit = Fit(pdf, xdata, objective=LogLikelihood)
fit_result = fit.execute()

self.assertAlmostEqual(fit_result.value(mu) / mean, 1.0)
self.assertAlmostEqual(fit_result.stdev(mu) / mean_stdev, 1.0)
self.assertAlmostEqual(fit_result.value(sig) / stdev, 1.0)
```

But we also know that the uncertainty in mu is given by sigma/sqrt(N), so if we find this in our `FitResults` we are happy.

If I use `ans.hess_inv` as the covariance matrix this seems to be in good agreement, but not exactly, for low N (5-100). However, as I increase N the uncertainty in mu seems to approach a certain asymptote and doesn't change any further.

After reading [this ](https://www.statlect.com/fundamentals-of-statistics/maximum-likelihood-covariance-matrix-estimation) I decided to implement the outer product of gradients (OPG) as specified there, and this given the exact result for all N. Comparing this outer product with `ans.hess_inv` we find the same; it is almost the same at small N but not at large N. It would therefore seem that the `L-BFGS-B` algorithm is unable to give a good estimate or, more likely, that it has something to do with the order of summation, differentiation, and matrix inversion that does not commute since the `L-BFGS-B` algorithm can only work with the scalar value returned by the LogLikelihoog function but to compute the OPG we need the gradient of the likelihood function at every individual point before summing and inverting.

I will therefore proceed to implement the OPG method to estimate the covariance matrix, and I will also add a `NotImplementedError` when trying to add weights to data in likelihood fitting since, well, that has not been implemented (https://arxiv.org/pdf/math/0505599.pdf).

**Comment by tBuLi** [comment]
Closed and 0.4.0 has been released :+1: .

Thanks for bringing this to my attention. Weights are left for another day.

**Comment by egede** [comment]
Fantastic work. I will test it out in the new year on a significantly more complex problem and report if I see any problems.
